engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/ROverseer.java

```diff
@@ -0,0 +1,153 @@
+            } else {
+                callback.accept(Stage.ABORT, "Timed out waiting for termination");
+            }
+        } else {
+            log.info("Overseer [{}] received abort, but process exited already", id);
+        }
+        return result;
+    }
+
+    @Override
+    public String getStdout() {
+        return stdout.toString();
+    }
+
+    private void writeLogs(String stage) {
+        try {
+            char[] buffer = new char[1024];
+            StringBuilder sb = new StringBuilder();
+            while (reader.ready()) {
+                int read = reader.read(buffer);
+                sb.append(buffer, 0, read);
+            }
+            if (sb.length() != 0) {
+                String delta = sb.toString();
+                log.info("STDOUT [{}]:\n{}", id, delta);
+                stdout.append(delta);
+                callback.accept(stage, delta);
+            }
+            // Do this after flushing the log to ensure that end of log doesn't get lost
+            if (!process.isAlive()) {
+                log.info("Overseer [{}] shutdown", id);
+                complete(process.exitValue());
+            }
+        } catch (IOException e) {
+            if (process.isAlive()) {
+                log.error("Overseer [{}] unable to read log on a live process: {}", id, e.getMessage());
+                callback.accept(stage, "=== Error reading log ===");
+                logFlush.cancel(false);
+            } else {
+                log.warn("Overseer [{}] shutdown, dead process, read log error: {}", id, e.getMessage());
+                complete(process.exitValue());
+            }
+        }
+    }
+
+    private void complete(int exitValue) {
+        executor.shutdown();
+        ExecutionOutcome outcome = (exitValue == 0)
+                ? new ExecutionOutcome(Stage.COMPLETED, null, stdout.toString())
+                : new ExecutionOutcome(Stage.EXECUTE, "Exit code " + exitValue, stdout.toString());
+        result.complete(outcome);
+    }
+
+    private boolean terminate() {
+        // TODO Can race with cancel?
+        writeLogs(Stage.ABORT);
+        boolean dead = waitForKill();
+        if (dead) {
+            log.info("Terminated [{}]", id);
+            executor.shutdownNow();
+            log.info("Shut down overseer for [{}]", id);
+        } else {
+            // We don't want to shut down executor in order to get some logs
+            log.error("Overseer [{}] failed to terminate analysis process", id);
+        }
+        return dead;
+    }
+
+    private boolean waitForKill() {
+        try {
+            // Will not kill whole process tree on windows. The fundamental problem here is that (unlike Unix)
+            // Windows doesn't maintain parent-child relationships between processes. A process can kill
+            // its own immediate children, but not 'grand-children' because it has no way of finding them.
+            return process.destroyForcibly().waitFor(30, TimeUnit.SECONDS);
```

Comment:
Probably it's better to make timeout configurable

Reply:
Good catch. I was planning to do so initially but overlooked it later on.
Added optional setting: `runtime.killTimeoutSec`